python/cuml/metrics/hinge_loss.pyx

```diff
@@ -0,0 +1,126 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cudf
+import cupy as cp
+from cuml.common.input_utils import determine_array_type
+from cuml.preprocessing import LabelEncoder, LabelBinarizer
+
+
+def hinge_loss(y_true,
+               pred_decision,
+               labels=None,
+               sample_weights=None) -> float:
+    """
+    Calculates non-regularized hinge loss. Adapted from scikit-learn hinge loss
+
+    Parameters
+    ----------
+    y_true: cuDF Series or cuPy array of shape (n_samples,)
+        True labels, consisting of labels for the classes.
+        In binary classification, the positive label must be
+        greater than negative class
+
+    pred_decision: cuDF DataFrame or cuPy array of shape (n_samples,) or \
+            (n_samples, n_classes)
+        Predicted decisions, as output by decision_function (floats)
+
+    labels: cuDF Series or cuPy array, default=None
+            In multiclass problems, this must include all class labels.
+
+    sample_weight: cupy array of shape (n_samples,), default=None
+                   Sample weights to be used for computing the average
+
+    Returns
+    -------
+    loss : float
+           The average hinge loss.
+    """
+
+    yt_type = determine_array_type(y_true)
+    pd_type = determine_array_type(pred_decision)
+    labels_type = determine_array_type(labels)
+
+    if yt_type not in ['cupy', 'numba', 'cudf']:
+        raise TypeError("y_true needs to be either a cuDF Series or \
+                        a cuPy/numba array.")
+
+    if pd_type not in ['cupy', 'numba', 'cudf']:
+        raise TypeError("pred_decision needs to be either a cuDF DataFrame or \
+                        a cuPy/numba array.")
+
+    if labels_type not in ['cupy', 'numba', 'cudf']:
+        raise TypeError("labels needs to be either a cuDF Series or \
+                        a cuPy/numba array.")
+
+    if y_true.shape[0] != pred_decision.shape[0]:
+        raise ValueError("y_true and pred_decision must have the same"
+                         " number of rows(found {} and {})".format(
+                             y_true.shape[0],
+                             pred_decision.shape[0]))
+
+    if sample_weights and sample_weights.shape[0] != y_true.shape[0]:
+        raise ValueError("y_true and sample_weights must have the same "
+                         "number of rows (found {} and {})".format(
+                             y_true.shape[0],
+                             sample_weights.shape[0]))
+
+    if not isinstance(labels, cudf.Series):
+        labels = cudf.Series(labels)
+
+    if not isinstance(y_true, cudf.Series):
+        y_true = cudf.Series(y_true)
+
+    y_true_unique = cp.unique(labels if labels is not None else y_true)
```

Comment:
Should these cupy functions or the broader `hinge_loss` function be wrapped to use RMM for allocations?

Reply:
Good catch. @Nanthini10 I mentioned this in an earlier review but you need to add `@cuml.internals.api_return_any()` to `hinge_loss`. Should look like this:
```python
@cuml.internals.api_return_any()
def hinge_loss(y_true,
               pred_decision,
               labels=None,
               sample_weights=None) -> float:
```
Otherwise the RMM allocators will not be used.

Reply:
@Nanthini10 Looks like you removed the needed decorator in commit https://github.com/rapidsai/cuml/pull/3409/commits/ae070f0045274a83c6b9f605bb60516eb1f74ac0

Reply:
@mdemoret now **that** is a nice decorator. Would it be right to say the general developer guidelines would be to wrap function declarations with that? If so, I think we can probably close https://github.com/rapidsai/cuml/issues/2485

Reply:
@mdemoret Yeah, I noticed yesterday that I removed it accidentally from the commit. I'm just wrapping up another issue, will finish this up. Should be a straightforward fix.

Reply:
Added it! Thanks, @beckernick. Let me know if this looks good!

Reply:
> @mdemoret now **that** is a nice decorator. Would it be right to say the general developer guidelines would be to wrap function declarations with that? If so, I think we can probably close #2485

Yes, all publicly facing functions should be wrapped in decorators. I have a section about how the decorators should be used in the [Estimator Guide](https://github.com/rapidsai/cuml/blob/branch-0.19/wiki/python/ESTIMATOR_GUIDE.md#option-2-manual-estimator-method-decoration). While that section is focused on Estimator functions, the same applies for non-Estimator functions (see the `non-Base` row in the table I linked above).

I think its safe to say #2485 can be closed because neither `@with_cupy_rmm` or `@rmm_cupy_ary` should be used anymore. They are built into the new decorators and Estimator Descriptors (see [here](https://github.com/rapidsai/cuml/blob/branch-0.19/wiki/python/ESTIMATOR_GUIDE.md#working-with-cumlarraydescriptor) for a guide on those)

Reply:
Love it. Clean, clear, and well documented. Closing that issue.